xdsl/dialects/mpi.py

```diff
@@ -0,0 +1,45 @@
+from __future__ import annotations
+
+from xdsl.ir import OpResult, SSAValue, Dialect
+from xdsl.ir import ParametrizedAttribute
+
+from xdsl.irdl import (irdl_op_definition, Operation, Operand, Annotated,
+                       irdl_attr_definition, AnyAttr)
+
+
+@irdl_attr_definition
+class MPIRequest(ParametrizedAttribute):
+    name = "MPI_Request"
+
+
+@irdl_attr_definition
+class MPIStatus(ParametrizedAttribute):
+    name = "MPI_Status"
+
+
+@irdl_op_definition
+class MPIRequestOp(Operation):
```

Comment:
In this code (from [here](https://rookiehpc.github.io/mpi/docs/mpi_request/index.html)): 

```C
int received;
MPI_Request request;
MPI_Irecv(&received, 1, MPI_INT, SENDER, 0, MPI_COMM_WORLD, &request);

// Do some other things while the underlying MPI_Recv progresses.
printf("MPI process %d issued the MPI_Irecv and obtained an MPI_Request.\n", my_rank);
 
// Wait for the MPI_Recv to complete.
printf("MPI process %d waits on that MPI_Request.\n", my_rank);
MPI_Wait(&request, MPI_STATUS_IGNORE);
printf("The MPI_Wait completed, which means the request (pointing to an MPI_Recv) is complete too (value received: %d).\n", received);
break;
```

This models the `MPI_Request request;` line, right? So it is pretty much just an allocation of some memory to write the MPI_Request into?

So essentially, this is just the thing that tells the MPI_Wait what to what for, right?

In that case, I don't quite understand why you have an operand here. I would have expected just a single result.

Reply:
Also, I am not quite sure about how working with MPI really works, but from the code above, it feels to me that the MPI_Irecv should have produced the request, since it is the first one to actually write something into it. If others can also write into it before MPI_Wait "consumes" it, then your design makes more sense.

Also, I am guessing that this is becoming a design discussion and that I am thinking of an "upward facing" design (as Paul described it), whereas this is more of a "downward facing" design.